src/go/rpk/pkg/cli/cmd/root.go

```diff
@@ -89,3 +104,85 @@ func Execute() {
 		os.Exit(1)
 	}
 }
+
+type pluginHandler interface {
+	lookPath(file string) (path string, ok bool)
+	exec(path string, args []string) error
+}
+
+// tryExecPlugin looks for a plugin, following the exact behavior of kubernetes:
+//
+//  - all dashes are turned to underscores
+//  - all "pieces" (non-flags) are joined with a dash
+//  - we prefer the longest command match
+//  - we search upward by piece until we run out of pieces
+//  - lastly, the command must be executable
+//
+// Dashes are only replaced into underscores during the path search; if an
+// argument hash a dash and is used as an argument to a plugin, the dash
+// remains.
+//
+// So,
+//
+//     rpk foo-bar baz boz fizz-buzz --flag
+//
+// is translated into searching and execing (with osPluginHandler), in order:
+//
+//     rpk-foo_bar-baz-boz-fizz_buzz (with args "--flag")
+//     rpk-foo_bar-baz-boz           (with args "fizz-buzz --flag")
+//     rpk-foo_bar-baz               (with args "boz fizz-buzz --flag")
+//     rpk-foo_bar                   (with args "baz boz fizz-buzz --flag")
+//
+// If a plugin is run, this returns the run error and true, otherwise this
+// returns false.
+func tryExecPlugin(h pluginHandler, args []string) (string, error) {
+	var pieces []string
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "-") { // found a flag, quit
+			break
+		}
+		pieces = append(pieces, strings.Replace(arg, "-", "_", -1))
```

Comment:
nit: `ReplaceAll` can be used instead: https://golang.org/pkg/strings/#ReplaceAll